### PR TITLE
iOS platform ODQ

### PR DESCRIPTION
**Pull request by joel-meng-nz**
Updated bindings of iOS, 
- Update MEGAError for iOS
- Add method to obtain `deadline` and `warningTs`
- Update `GlobalDelegate` `onEvent` documentation.

### CODE REVIEW

File: bindings/ios/MEGASdk.h [hunk 1]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
  */
 - (void)killSession:(uint64_t)sessionHandle;
 
+/**
+ * @brief Returns the deadline to remedy the storage overquota situation
+ *
+ * This value is valid only when MegaApi::getUserData has been called after
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5407]: Better use the binding methods / enums
In example: [MegaSdk getUserData], 
MegaListener/MegaGlobalListener::onEvent ---> [MEGADelegate/MEGAGlobalDelegate onEvent:event:]

STORAGE_STATE_PAYWALL ---> StorageStatePaywall
MegaEvent::EVENT_STORAGE ---> EventStorage

Etc

File: bindings/ios/MEGASdk.h [hunk 2]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
...
+ * The value will become invalid once the state of storage changes.
+ *
+ * @return Timestamp representing the deadline to remedy the overquota
+*/
+- (uint64_t)getOverquotaDeadlineTs;
+
+/**
+ * @brief Returns when the user was warned about overquota state
+ *
+ * This value is valid only when MegaApi::getUserData has been called after
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5419]: Same

File: bindings/ios/MEGASdk.h [hunk 3]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
...
+ * @brief Returns the deadline to remedy the storage overquota situation
+ *
+ * This value is valid only when MegaApi::getUserData has been called after
+ * receiving a callback MegaListener/MegaGlobalListener::onEvent of type
+ * MegaEvent::EVENT_STORAGE, reporting STORAGE_STATE_PAYWALL.
+ * The value will become invalid once the state of storage changes.
+ *
+ * @return Timestamp representing the deadline to remedy the overquota
+*/
+- (uint64_t)getOverquotaDeadlineTs;
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5414]: Please see this: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html

Use “get” only for methods that return objects and values indirectly.

File: bindings/ios/MEGASdk.h [hunk 4]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
...
+ * This value is valid only when MegaApi::getUserData has been called after
+ * receiving a callback MegaListener/MegaGlobalListener::onEvent of type
+ * MegaEvent::EVENT_STORAGE, reporting STORAGE_STATE_PAYWALL.
+ * The value will become invalid once the state of storage changes.
+ *
+ * You take the ownership of the returned value.
+ *
+ * @return An array of `NSNumber` with the timestamp corresponding to each warning
+*/
+-(nonnull NSArray<NSNumber *> *)getOverquotaWarningsTs;
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5428]: Same as `getOverquotaDeadlineTs`

File: bindings/ios/MEGASdk.mm [hunk 5]
```diff
@@ -1497,6 +1497,26 @@ - (void)killSession:(uint64_t)sessionHandle {
     self.megaApi->killSession(sessionHandle);
 }
 
+- (uint64_t)getOverquotaDeadlineTs {
+    return self.megaApi->getOverquotaDeadlineTs();
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.mm:1501]: We can return here the ts directly or `NSDate` if you prefer.

File: bindings/ios/MEGASdk.mm [hunk 6]
```diff
@@ -1497,6 +1497,26 @@ - (void)killSession:(uint64_t)sessionHandle {
...
+- (nonnull NSArray<NSNumber *> *)getOverquotaWarningsTs {
+    MegaIntegerList *warningTimestampList = self.megaApi->getOverquotaWarningsTs();
+
+    int sizeOfWarningTimestamps = warningTimestampList->size();
+    if (sizeOfWarningTimestamps == 0) {
+        return [[NSMutableArray alloc] initWithCapacity:0];
+    }
+
+    NSMutableArray *warningTimestampArray = [[NSMutableArray alloc] initWithCapacity:sizeOfWarningTimestamps];
+    for(int i = 0; i < sizeOfWarningTimestamps; i ++) {
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.mm:1513]:
```suggestion
    for(int i = 0; i < sizeOfWarningTimestamps; i++) {
```

File: bindings/ios/MEGASdk.mm [hunk 7]
```diff
@@ -1497,6 +1497,26 @@ - (void)killSession:(uint64_t)sessionHandle {
     self.megaApi->killSession(sessionHandle);
 }
 
+- (uint64_t)getOverquotaDeadlineTs {
+    return self.megaApi->getOverquotaDeadlineTs();
+}
+
+- (nonnull NSArray<NSNumber *> *)getOverquotaWarningsTs {
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.mm:1504]: Same here, maybe it is better return an array of `NSDate`

File: bindings/ios/MEGASdk.h (hunk 1 above)
Comment (joel-meng-nz) [reply bindings/ios/MEGASdk.h:5407]: Got it

File: bindings/ios/MEGAGlobalDelegate.h [hunk 8]
```diff
@@ -165,6 +165,11 @@ NS_ASSUME_NONNULL_BEGIN
  *     After calling it, this callback will be called again with the corresponding
  *     state if there is really a change.
  *
+ *     - StorageStatePaywall = 4
+ *     The account has been full for a long time. Now most of actions are disallowed.
+ *     You will need to call [MegaSdk getUserData] before retrieving the overquota deadline/warnings
```
Comment (simonwng) [inline bindings/ios/MEGAGlobalDelegate.h:170]: MEGASdk

File: bindings/ios/MEGASdk.h [hunk 9]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
  */
 - (void)killSession:(uint64_t)sessionHandle;
 
+/**
+ * @brief Returns the deadline to remedy the storage overquota situation
+ *
+ * This value is valid only when [MegaSdk getUserData] has been called after
+ * receiving a callback [MegaDelegate/MegaGlobalDelegate onEvent:event] of type
```
Comment (simonwng) [inline bindings/ios/MEGASdk.h:5408]: MegaGlobalDelegate is enough here, we don't need to expose sdk class names or interfaces

File: bindings/ios/MEGASdk.h [hunk 10]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
  */
 - (void)killSession:(uint64_t)sessionHandle;
 
+/**
+ * @brief Returns the deadline to remedy the storage overquota situation
+ *
+ * This value is valid only when [MegaSdk getUserData] has been called after
```
Comment (simonwng) [inline bindings/ios/MEGASdk.h:5407]: MEGASdk

File: bindings/ios/MEGASdk.mm [hunk 11]
```diff
@@ -1497,6 +1497,26 @@ - (void)killSession:(uint64_t)sessionHandle {
...
+- (nonnull NSDate *)overquotaDeadlineDate {
+    return [[NSDate alloc] initWithTimeIntervalSince1970:self.megaApi->getOverquotaDeadlineTs()];
+}
+
+- (nonnull NSArray<NSDate *> *)overquotaWarningDateList {
+    MegaIntegerList *warningTimeIntervalList = self.megaApi->getOverquotaWarningsTs();
+
+    int sizeOfWarningTimestamps = warningTimeIntervalList->size();
+    if (sizeOfWarningTimestamps == 0) {
+        return [[NSMutableArray alloc] initWithCapacity:0];
```
Comment (simonwng) [inline bindings/ios/MEGASdk.mm:1509]: NSArray? I don't think we need `NSMutableArray` here

Comment (joel-meng-nz) [reply bindings/ios/MEGASdk.mm:1509]: Yep, will fix this. But the thing is `Jenkins` build failed and I have no idea, what can I do to fix this? Ask someone who is in SDK to figure out?

Comment (simonwng) [reply bindings/ios/MEGASdk.mm:1509]: you can add a new commit to check how it goes, I don't think it is the problem in this PR

Comment (simonwng) [inline bindings/ios/MEGASdk.mm:1509]: we don't need to check 0 here, the for loop can handle 0 case. make the code simpler 😉

Comment (joel-meng-nz) [reply bindings/ios/MEGASdk.mm:1509]: 👍

File: bindings/ios/MEGASdk.h [hunk 12]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
...
+ * @brief Returns the deadline to remedy the storage overquota situation
+ *
+ * This value is valid only when [MEGASdk getUserData] has been called after
+ * receiving a callback [MEGAGlobalDelegate onEvent:event] of type
+ * EventStorage, reporting StorageStatePaywall.
+ * The value will become invalid once the state of storage changes.
+ *
+ * @return `NSDate` instance representing the deadline to remedy the overquota
+*/
+- (nonnull NSDate *)overquotaDeadlineDate;
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5414]: I think we can remove `nonnull` here. We have `NS_ASSUME_NONNULL_BEGIN` at the beginning of the file

File: bindings/ios/MEGASdk.h [hunk 13]
```diff
@@ -5400,6 +5401,32 @@ typedef NS_ENUM(NSInteger, AffiliateType) {
...
+ * This value is valid only when [MEGASdk getUserData] has been called after
+ * receiving a callback [MEGAGlobalDelegate onEvent:event] of type
+ * EventStorage, reporting StorageStatePaywall.
+ * The value will become invalid once the state of storage changes.
+ *
+ * You take the ownership of the returned value.
+ *
+ * @return An array of `NSDate` with the timestamp corresponding to each warning
+*/
+-(nonnull NSArray<NSDate *> *)overquotaWarningDateList;
```
Comment (jnavarrom) [inline bindings/ios/MEGASdk.h:5428]: same here

File: bindings/ios/MEGASdk.h (hunk 12 above)
Comment (simonwng) [reply bindings/ios/MEGASdk.h:5414]: yup, nonnull by default

**Comment by jnavarrom** [comment]
We can merge it

**Pull request merged**